### PR TITLE
Resource Based Routing

### PR DESCRIPTION
**Pull request by laljikanjareeya**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/nodejs-spanner/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> 🦕

### CODE REVIEW

File: src/instance.ts [hunk 1]
```diff
@@ -732,6 +733,57 @@ class Instance extends common.GrpcServiceObject {
...
+      (err, resp) => {
+        if (err) {
+          callback!(err);
+          return;
+        }
+        callback!(null, resp!.endpointUris);
+      }
+    );
+  }
+
```
Comment (laljikanjareeya) [inline src/instance.ts]: probably this method will be removed and will use existing method `getMetadata`, once @AVaksman expose the configuration part in that method.

**Comment by laljikanjareeya** [comment]
@jiren and @AVaksman your inputs are very much appreciate.

File: src/index.ts [hunk 2]
```diff
@@ -771,20 +775,78 @@ class Spanner extends GrpcService {
...
+              process.emitWarning(
+                'instance permisssion must be added to use resource based routing.'
+              );
+            } else {
+              callback(err);
+              return;
+            }
+          }
+          const options = Object.assign({}, this.options);
+          if (endPointUris && endPointUris!.length > 0) {
```
Comment (AVaksman) [inline src/index.ts]: The empty array (`endpointUris` might be an empty array) is a [`truthy`](https://developer.mozilla.org/en-US/docs/Glossary/Truthy) value, not necessary to check for it.
0 is a [`falsy`](https://developer.mozilla.org/en-US/docs/Glossary/Falsy) value.
I think this condition may be replaced with just
```ts
if (endpointUris.length)
```

**Comment by AVaksman** [comment]
```ts
it('should continue if do not have permission for instance', done => {
      sandbox.stub(process, 'emitWarning');
      const expectedWarning =
        'spanner.instances.get permisssion must be added to use resource based routing.';

      const error = new Error('Error.') as ServiceError;
      error.code = 7;
      const instanceId = 'instance-id';
      spanner.options.enableResourceBasedRouting = true;

      spanner.instance(instanceId).getInstanceEndPointUris = callback => {
        callback!(error);
      };
      asAny(CONFIG).instanceId = instanceId;

      spanner.prepareGapicRequest_(CONFIG, err => {
        assert.ifError(err);
        assert.ok(
          (process.emitWarning as sinon.SinonStub).calledWith(expectedWarning)
        );
        done();
      });
    });
```

```ts
  getGaxClient_(config, callback) {
    let clientName = config.client;
    if (
      clientName !== 'SpannerClient' ||
      !this.options.enableResourceBasedRouting
    ) {
      if (!this.clients_.has(clientName)) {
        this.setSpannerClient(clientName, config, this.options);
        // this.clients_.set(clientName, new gapic.v1[clientName](this.options));
      }
      const gaxClient = this.clients_.get(clientName)!;
      callback(null, gaxClient);
      return;
    }

    if (!config.instanceId) {
      const error = new Error('instanceId is required.');
      callback(error);
      return;
    }
    const instanceId = config.instanceId;
    clientName = `${clientName}-${instanceId}`;

    if (!this.clients_.has(clientName)) {
      this.instances_
        .get(instanceId)!
        .getInstanceEndPointUris((err, endPointUris) => {
          if (err) {
            if (err.code === 7) {
              process.emitWarning(
                'spanner.instances.get permisssion must be added to use resource based routing.'
              );
              this.setSpannerClient(clientName, config, this.options);
              callback(null, this.clients_.get(clientName)!);
              return;
            } else {
              callback(err);
              return;
            }
          }
          const options = Object.assign({}, this.options);
          if (endPointUris!.length) {
            options.apiEndpoint = endPointUris![0];
          }
          this.setSpannerClient(clientName, config, options);
          // this.clients_.set(clientName, new gapic.v1[config.client](options));
          callback(null, this.clients_.get(clientName)!);
        });
    } else {
      callback(null, this.clients_.get(clientName)!);
    }
  }

  setSpannerClient(clientName, config, options) {
    this.clients_.set(clientName, new gapic.v1[config.client](options));
  }
```

**Pull request closed without merging**